src/adapter/src/catalog/transact.rs

```diff
@@ -2262,6 +2268,137 @@ impl Catalog {
+                    required: Duration::from_millis(used_rh.comparable_timestamp().into()),
+                    found: Duration::from_millis(index_rh.comparable_timestamp().into()),
+                    name: catalog
+                        .resolve_full_name(used_by_index.name(), None)
+                        .to_string(),
+                });
+            }
+        }
+
+        Ok(())
+    }
+
+    /// For the Index at `id` that is on a View, return the index ids it directly depends on.
+    ///
+    /// # Panics
+    /// Panics if `id` is not an Index on a View.
+    fn get_view_index_uses(catalog: &CatalogState, id: &GlobalId) -> BTreeSet<GlobalId> {
+        let entry = catalog.get_entry(id);
+        let index = entry.index().expect("must be an index");
+        let mut index_ids = BTreeSet::new();
+        let mut views_to_check = VecDeque::from([index.on]);
+        while let Some(view_id) = views_to_check.pop_front() {
+            let view = catalog.get_entry(&view_id);
+            assert!(view.is_view());
+            // Views can use other views, need to check those too.
+            for uses_id in view.uses() {
+                match catalog.get_entry(&uses_id).item().typ() {
+                    CatalogItemType::View => {
+                        views_to_check.push_back(uses_id);
+                    }
+                    _ => {}
+                }
+            }
+            // Indexes are used by views: add those to the result set.
+            for used_by_id in view.used_by() {
+                match catalog.get_entry(used_by_id).item().typ() {
+                    CatalogItemType::Index => {
+                        index_ids.insert(*used_by_id);
+                    }
+                    _ => {}
+                }
+            }
+        }
+        index_ids.remove(id);
+        index_ids
+    }
+
+    /// For the Index at `id` that is on a View, return the index ids that use it.
+    ///
+    /// # Panics
+    /// Panics if `id` is not an Index on a View.
+    fn get_view_index_used_by(catalog: &CatalogState, id: &GlobalId) -> BTreeSet<GlobalId> {
+        let entry = catalog.get_entry(id);
+        let index = entry.index().expect("must be an index");
+        let mut index_ids = BTreeSet::new();
+        let mut views_to_check = VecDeque::from([index.on]);
+        while let Some(view_id) = views_to_check.pop_front() {
+            let view = catalog.get_entry(&view_id);
+            assert!(view.is_view());
+            for used_by_id in view.used_by() {
+                match catalog.get_entry(used_by_id).item().typ() {
+                    CatalogItemType::View => {
+                        views_to_check.push_back(*used_by_id);
+                    }
+                    CatalogItemType::Index => {
+                        index_ids.insert(*used_by_id);
+                    }
+                    _ => {}
+                }
+            }
+        }
+        index_ids.remove(id);
+        index_ids
+    }
```

Comment:
Tracking seen views/indexes might prevent duplicate checks and speed these functions up.

Reply:
These functions don't recurse, and it's not possible to have circular dependencies in SQL objects, so I don't think any work duplication is occurring.

Reply:
The performance difference is probably negligible if any, but I _think_ you can have a scenario like

```
CREATE VIEW a AS SELECT 1;
CREATE VIEW b AS SELECT * FROM a;
CREATE VIEW c AS SELECT * FROM a, b;
```

If we start with `c`, then our queue might look like
1. `[c]`
2. process(c) -> `[a, b]`
3. process(a) -> `[b]`
4. process(b) -> `[a]`
5. process(a) -> `[]`

which means we would end up processing `a` twice.

Reply:
Agree that this is possible and probably negligible!